File: day_1/OneSpec.hs
import Test.Hspec
import One

main :: IO ()
main = hspec $ do
    describe "finalFloor" $ do
        it "returns 0 on an empty string" $ do
            finalFloor "" `shouldBe` 0

        it "returns -1 for one right paren" $ do
            finalFloor ")" `shouldBe` (-1)

        it "returns 1 for one left paren" $ do
            finalFloor "(" `shouldBe` 1

        it "returns 0 for '()'" $ do
            finalFloor "()" `shouldBe` 0

        it "returns 3 for '(()(()('" $ do
            finalFloor "(()(()(" `shouldBe` 3

    describe "basementChar" $ do
        it "returns 1 on ')'" $ do
            basementChar ")" `shouldBe` 1

        it "returns 5 on '()())'" $ do
            basementChar "()())" `shouldBe` 5

File: 2015/lua/day_1/one_spec.lua
local one = require("one")

describe("final_floor", function()
    it("returns 3 for the string '((('", function()
        assert.are.equal(one.final_floor("((("), 3)
    end)

    it("returns -3 for the string ')))'", function()
        assert.are.equal(one.final_floor(")))"), (-3))
    end)

    it("returns 0 for the string '((()))'", function()
        assert.are.equal(one.final_floor("((()))"), 0)
    end)

    it("errors if one of the input characters is not a paren", function()
        assert.has_error(
            function() one.final_floor("()()x") end,
            "Not a paren: x"
        )
    end)
end)

describe("basement_char", function()
    it("returns how many chars it takes to get to a negative floor", function()
        assert.are.equal(one.basement_char("(((()))))"), 9)
    end)
end)
